samza-core/src/main/java/org/apache/samza/execution/LocalJobPlanner.java

```diff
@@ -119,21 +137,79 @@ private void createStreams(String planId, List<StreamSpec> intStreams, StreamMan
       return;
     }
 
-    DistributedLockWithState lockWithState = coordinationUtils.getLockWithState(planId);
+    // If BATCH, then need to create new intermediate streams every run.
+    // planId does not change every run and hence, need to use runid
+    // as the lockId to create a new lock with state each run
+    // to create new streams each run.
+    // If run.id is null, defaults to old behavior of using planId
+    boolean isAppModeBatch = new ApplicationConfig(userConfig).getAppMode() == ApplicationConfig.ApplicationMode.BATCH;
+    String lockId = planId;
+    if (isAppModeBatch && runId != null) {
+      lockId = runId;
+    }
     try {
-      // check if the processor needs to go through leader election and stream creation
-      if (lockWithState.lockIfNotSet(1000, TimeUnit.MILLISECONDS)) {
-        LOG.info("lock acquired for streams creation by " + uid);
+      MetadataStore metadataStore = getMetadataStore();
+      DistributedLock distributedLock = coordinationUtils.getLock(lockId);
+      if (distributedLock == null || metadataStore == null) {
+        LOG.warn("Processor {} failed to create utils. Each processor will attempt to create streams.", uid);
+        // each application process will try creating the streams, which requires stream creation to be idempotent
         streamManager.createStreams(intStreams);
-        lockWithState.unlockAndSet();
-      } else {
-        LOG.info("Processor {} did not obtain the lock for streams creation. They must've been created by another processor.", uid);
+        return;
       }
-    } catch (TimeoutException e) {
-      String msg = String.format("Processor {} failed to get the lock for stream initialization", uid);
-      throw new SamzaException(msg, e);
+      //Start timer for timeout
+      long startTime = System.currentTimeMillis();
+      long lockTimeout = TimeUnit.MILLISECONDS.convert(CoordinationConstants.LOCK_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+
+      // If "stream created state" exists in store then skip stream creation
+      // Else acquire lock, create streams, set state in store and unlock
+      // Checking for state before acquiring lock to prevent all processors from acquiring lock
+      // In a while loop so that if two processors check state simultaneously then
+      // to make sure the processor not acquiring the lock
+      // does not die of timeout exception and comes back and checks for state and proceeds
+      while ((System.currentTimeMillis() - startTime) < lockTimeout) {
```

Comment:
Can we modularize the following code? It is hard to read through especially with try and catch and throw statements everywhere.